forgerock-openbanking-uk-aspsp-rs/forgerock-openbanking-uk-aspsp-rs-mock-store/forgerock-openbanking-uk-aspsp-rs-mock-store-server/src/main/java/com/forgerock/openbanking/aspsp/rs/store/repository/migration/v3_1_6/MongoDbPaymentsChangeLog.java

```diff
@@ -24,83 +24,153 @@
+        elapsedTime.stop();
 
-        CloseableIterator<FRFilePaymentSubmission2> frFilePaymentSubmissions = getLegacyDocuments(mongoTemplate, FRFilePaymentSubmission2.class);
-        frFilePaymentSubmissions.forEachRemaining(d -> migrate(mongoTemplate, d, toFRFilePaymentSubmission(d)));
+        log.info("Upgraded {} documents in {} seconds.", docsUpdated, elapsedTime.getTotalTimeSeconds());
+        log.info("-----------------------------------------------------------------------");
+        log.info("Finished migrating Payments API data from v3.1.6 to v3.1.6 patch");
     }
 
-    private void migrateInternationalPayments(MongoTemplate mongoTemplate) {
-        CloseableIterator<FRInternationalConsent2> frInternationalConsents = getLegacyDocuments(mongoTemplate, FRInternationalConsent2.class);
-        frInternationalConsents.forEachRemaining(d -> migrate(mongoTemplate, d, toFRInternationalConsent(d)));
+    /*
+     * Entry method to execute the upgrade bulk operation
+     * @param mongoTemplate
+     * @param legacyClazz collection class
+     * @param writeRiskParentField field name to get the write object that contains the risk object related with the collection class
+     * @param <T> generic parameter to conforms legacyClazz object
+     */
+    private <T> int upgrade(MongoTemplate mongoTemplate, Class<T> legacyClazz, String writeRiskParentField) {
+        // execution time control - start
+        StopWatch elapsedTime = start(legacyClazz);
+
+        // number of docs updated
+        int docsUpdated = 0;
+        // prepare the json path to build the criteria filter
+        String jsonPathCriteria = new StringBuffer(writeRiskParentField)
+                .append(SEPARATOR).append(RISK)
+                .append(SEPARATOR).append(DELIVERY_ADDRESS)
+                .append(SEPARATOR).append(COUNTRY_SUB_DIVISION).toString();
+
+        // Criteria to query only documents where countrySubDivision not null and obVersion is 'v3_1_6'
+        // { "${writeRiskParentField}.risk.deliveryAddress.countrySubDivision" : { "$ne" : null } }
+        Criteria queryCriteria = new Criteria(jsonPathCriteria)
+                .ne(null);
+
+        BulkOperations bulkOperations = mongoTemplate.bulkOps(BulkOperations.BulkMode.UNORDERED, legacyClazz);
+        // get the documents filter by criteria
+        Query query = new Query().addCriteria(queryCriteria);
+        Update update = new Update();
+        List<Pair<Query, Update>> updates = new ArrayList<>();
+        mongoTemplate.find(query, legacyClazz).forEach(
+                legacyClassDocInstance -> {
+                        // getting the id and the first element of countrySubDivision from the document
+                        String id = ((LegacyCountrySubDivision)legacyClassDocInstance).getDocumentId();
+                        String division = ((LegacyCountrySubDivision)legacyClassDocInstance).getCountrySubDivision();
```

Comment:
Looks good. I know this was my suggestion, but you have to admit it's a lot easier to read! 😄